spring-aot/src/main/java/org/springframework/context/bootstrap/generator/bean/descriptor/InitializationCallbacksSupplier.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.context.bootstrap.generator.bean.descriptor;
+
+import javax.annotation.PostConstruct;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import com.squareup.javapoet.CodeBlock;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.support.RootBeanDefinition;
+import org.springframework.context.bootstrap.generator.bean.descriptor.BeanInstanceDescriptor.InitializationCallback;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * Provide the {@link InitializationCallback initialization (post construct) callbacks} of a given class.
+ *
+ * @author Christoph Strobl
+ */
+class InitializationCallbacksSupplier {
+
+	private static final Field EXTERNALLY_MANAGED_INIT_METHODS;
+
+	static {
+
+		EXTERNALLY_MANAGED_INIT_METHODS = ReflectionUtils.findField(RootBeanDefinition.class, "externallyManagedInitMethods");
+		ReflectionUtils.makeAccessible(EXTERNALLY_MANAGED_INIT_METHODS);
+	}
+
+	List<InitializationCallback> detectInstanceCallbacks(BeanDefinition beanDefinition) {
+
+		if(!(beanDefinition instanceof RootBeanDefinition)) {
+			return Collections.emptyList();
+		}
+
+		RootBeanDefinition rootBeanDefinition = (RootBeanDefinition) beanDefinition;
+		Object field = ReflectionUtils.getField(EXTERNALLY_MANAGED_INIT_METHODS, rootBeanDefinition);
+
+		if (ObjectUtils.isEmpty(field)) {
+			return Collections.emptyList();
+		}
+
+		List<InitializationCallback> callbacks = new ArrayList<>();
+		Class<?> targetType = beanDefinition.getResolvableType().toClass();
+		for (Object initMethodName : (Iterable<? extends Object>) field) {
```

Comment:
We should rather put everything in a `LinkedHashSet` and iterate over it, to avoid cases where something is both detected by an initMethodName, and also annotated.